File: packages/contracts/src/audit.ts
import { z } from "zod";

import { approvalDecisionSchema, executionStatusSchema } from "./action-plan";
import { cuaActionRequestSchema, cuaActionResultSchema, cuaWindowStateSchema } from "./cua";
import { resolvedIntentSchema } from "./intent";
import { selectedReferentSchema } from "./referent";
import { surfaceSnapshotSchema } from "./surface";

// One audit-trail entry for the "select context" step of the core loop: the
// user pointed (referent) at a surface (snapshot), which a session/action then
// used. Persisting it makes the selection replayable — which surface did the
// user pick, how confident was perception, and which action consumed it
// (#23, epic #6).

export const surfaceSelectionRecordSchema = z.object({
  // The selected referent: id, source, confidence.
  referent: selectedReferentSchema,
  // The surface metadata as it was at selection time.
  surface: surfaceSnapshotSchema,
  // The supervision session the selection happened in. Always present — a
  // selection only has meaning inside a session.
  sessionId: z.string().min(1),
  // The action that consumed the selection. Optional because selection precedes
  // planning in the loop (select -> speak -> plan -> approve -> act); it is set
  // once an action claims the referent.
  actionId: z.string().min(1).optional(),
  // When the user made the selection, ISO 8601. Captured by the caller at
  // selection time, not when the record is persisted.
  selectedAt: z.string().datetime(),
});
export type SurfaceSelectionRecord = z.infer<typeof surfaceSelectionRecordSchema>;

// Validate an untrusted selection record at a boundary (IPC, persistence).
export function safeParseSurfaceSelectionRecord(
  input: unknown,
): z.SafeParseReturnType<unknown, SurfaceSelectionRecord> {
  return surfaceSelectionRecordSchema.safeParse(input);
}

const auditEventBaseSchema = z.object({
  sessionId: z.string().min(1),
  actionId: z.string().min(1),
  recordedAt: z.string().datetime(),
});

export const supervisionAuditEventSchema = z
  .discriminatedUnion("kind", [
    auditEventBaseSchema.extend({
      kind: z.literal("intent_created"),
      intent: resolvedIntentSchema,
    }),
    auditEventBaseSchema.extend({
      kind: z.literal("approval_decided"),
      approval: approvalDecisionSchema,
    }),
    auditEventBaseSchema.extend({
      kind: z.literal("cua_state_captured"),
      phase: z.enum(["pre", "post"]),
      stepId: z.string().min(1),
      state: cuaWindowStateSchema,
    }),
    auditEventBaseSchema.extend({
      kind: z.literal("cua_call"),
      stepId: z.string().min(1),
      request: cuaActionRequestSchema,
      result: cuaActionResultSchema,
    }),
    auditEventBaseSchema.extend({
      kind: z.literal("execution_finished"),
      status: executionStatusSchema,
      result: cuaActionResultSchema.optional(),
    }),
  ])
  .refine(
    (event) => event.kind !== "approval_decided" || event.approval.actionId === event.actionId,
    {
      message: "approval actionId must match audit actionId",
    },
  );
export type SupervisionAuditEvent = z.infer<typeof supervisionAuditEventSchema>;

export function safeParseSupervisionAuditEvent(
  input: unknown,
): z.SafeParseReturnType<unknown, SupervisionAuditEvent> {
  return supervisionAuditEventSchema.safeParse(input);
}

File: apps/sidecar/DirectorSidecar/DirectorSidecar/Contracts/Audit.swift
//
//  Audit.swift
//  DirectorSidecar
//
//  Port of @handsoff/contracts audit.ts — `supervisionAuditEventSchema` (the Intention Log
//  entry, discriminated on `kind`) and `surfaceSelectionRecordSchema`. Faithful to all six
//  audit kinds, including the autonomous-loop per-call `tool_call` record (U3) that carries
//  transcript + bound referent + derived risk + approval provenance.
//
//  Faithful refinements (mirror the zod `.refine`s, enforced as throwing decodes):
//   - every event requires `sessionId` + `actionId` (a kind with no action link fails).
//   - `approval_decided`: the embedded approval's `actionId` must equal the event's `actionId`.
//   - `tool_call`: an unknown driver tool name fails (DriverTool enum decode).
//

import Foundation

extension Contracts {
    /// `surfaceSelectionRecordSchema`: the "select context" audit step — the user pointed
    /// (referent) at a surface (snapshot), which an action later consumed. `actionId` is
    /// optional (selection precedes planning); `sessionId` is always present.
    struct SurfaceSelectionRecord: Codable, Sendable, Equatable {
        let referent: SelectedReferent
        let surface: SurfaceSnapshot
        let sessionId: String
        let actionId: String?
        let selectedAt: String
    }

    /// `supervisionAuditEventSchema`. The shared base (`sessionId`/`actionId`/`recordedAt`)
    /// is carried on every case.
    enum SupervisionAuditEvent: Decodable, Sendable, Equatable {
        case intentCreated(Base, intent: ResolvedIntent)
        case approvalDecided(Base, approval: ApprovalDecision)
        case cuaStateCaptured(Base, phase: CapturePhase, stepId: String, state: CuaWindowState)
        case cuaCall(Base, stepId: String, request: CuaActionRequest, result: CuaActionResult)
        case toolCall(Base, ToolCall)
        case executionFinished(Base, status: ExecutionStatus, result: CuaActionResult?)

        /// The shared audit base every event carries.
        struct Base: Sendable, Equatable {
            let sessionId: String
            let actionId: String
            let recordedAt: String
        }

        enum CapturePhase: String, Decodable, Sendable { case pre, post }

        /// How a per-call action was gated: `auto` (under threshold), or a human decision.
        enum ToolCallApproval: String, Decodable, Sendable { case auto, approved, rejected }

        /// The `tool_call` payload — the full provenance the Intention Log replays.
        struct ToolCall: Sendable, Equatable {
            let transcript: String
            let referent: SelectedReferent?    // null for referent-less calls (get_window_state)
            let tool: DriverTool
            let target: ToolCallTarget?
            let risk: RiskLevel
            let approval: ToolCallApproval
            let result: CuaActionResult
        }

        private enum Key: String, CodingKey {
            case kind, sessionId, actionId, recordedAt
            case intent, approval, phase, stepId, state, request, result, status
            case transcript, referent, tool, target, risk
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: Key.self)
            let base = Base(
                sessionId: try c.decode(String.self, forKey: .sessionId),
                actionId: try c.decode(String.self, forKey: .actionId),
                recordedAt: try c.decode(String.self, forKey: .recordedAt))

            switch try c.decode(String.self, forKey: .kind) {
            case "intent_created":
                self = .intentCreated(base,
                                      intent: try c.decode(ResolvedIntent.self, forKey: .intent))
            case "approval_decided":
                let approval = try c.decode(ApprovalDecision.self, forKey: .approval)
                guard approval.actionId == base.actionId else {
                    throw DecodingError.dataCorruptedError(
                        forKey: .approval, in: c,
                        debugDescription: "approval actionId must match audit actionId")
                }
                self = .approvalDecided(base, approval: approval)
            case "cua_state_captured":
                self = .cuaStateCaptured(
                    base,
                    phase: try c.decode(CapturePhase.self, forKey: .phase),
                    stepId: try c.decode(String.self, forKey: .stepId),
                    state: try c.decode(CuaWindowState.self, forKey: .state))
            case "cua_call":
                self = .cuaCall(
                    base,
                    stepId: try c.decode(String.self, forKey: .stepId),
                    request: try c.decode(CuaActionRequest.self, forKey: .request),
                    result: try c.decode(CuaActionResult.self, forKey: .result))
            case "tool_call":
                self = .toolCall(base, ToolCall(
                    transcript: try c.decode(String.self, forKey: .transcript),
                    referent: try c.decodeIfPresent(SelectedReferent.self, forKey: .referent),
                    tool: try c.decode(DriverTool.self, forKey: .tool),
                    target: try c.decodeIfPresent(ToolCallTarget.self, forKey: .target),
                    risk: try c.decode(RiskLevel.self, forKey: .risk),
                    approval: try c.decode(ToolCallApproval.self, forKey: .approval),
                    result: try c.decode(CuaActionResult.self, forKey: .result)))
            case "execution_finished":
                self = .executionFinished(
                    base,
                    status: try c.decode(ExecutionStatus.self, forKey: .status),
                    result: try c.decodeIfPresent(CuaActionResult.self, forKey: .result))
            case let other:
                throw DecodingError.dataCorruptedError(
                    forKey: .kind, in: c,
                    debugDescription: "Unknown audit event kind: \(other)")
            }
        }
    }
}
